File: DataEngineering/08_Data_Governance_and_Quality/projects/python/05_schema_validation_gate/main.py
import json
from pydantic import BaseModel, ValidationError, Field, EmailStr
from datetime import datetime
from typing import Optional

# Define the expected schema for incoming data using Pydantic.
# This acts as a strict schema validation gate for data entering the pipeline.
class UserEvent(BaseModel):
    event_id: str = Field(..., min_length=10, description="Unique identifier for the event")
    user_id: int = Field(..., gt=0)
    email: EmailStr
    event_timestamp: datetime
    is_active: bool
    platform: Optional[str] = "web"

def process_event(raw_payload: str):
    """
    Validates a raw JSON string against the UserEvent schema.
    If valid, proceeds with downstream processing.
    If invalid, routes to a Dead Letter Queue (DLQ).
    """
    print(f"\nProcessing payload:\n{raw_payload}")
    try:
        # Parse and validate the schema
        event_dict = json.loads(raw_payload)
        validated_event = UserEvent(**event_dict)
        
        print("✅ Schema Validation Passed!")
        # Accessing typed and validated fields
        print(f"Proceeding to ingest event: {validated_event.event_id} from user {validated_event.user_id}")
        
    except json.JSONDecodeError:
        print("❌ Invalid JSON format. Routing to DLQ...")
    except ValidationError as e:
        print("❌ Schema Validation Failed. Routing to DLQ...")
        print(f"Validation Errors:\n{e.json()}")

def main():
    # Good Payload: Matches the schema perfectly
    good_payload = json.dumps({
        "event_id": "evt_123456789",
        "user_id": 42,
        "email": "alice@company.com",
        "event_timestamp": "2023-10-25T14:30:00Z",
        "is_active": True
    })

    # Bad Payload 1: Missing required field (event_timestamp), invalid email, user_id is negative
    bad_payload = json.dumps({
        "event_id": "evt_987", # Too short
        "user_id": -5,
        "email": "not-an-email",
        "is_active": "yes" # Should be boolean, but Pydantic might try to coerce. Let's see. 'yes' can't be coerced to bool easily.
    })

    process_event(good_payload)
    process_event(bad_payload)

if __name__ == "__main__":
    main()

File: DataEngineering/08_Data_Governance_and_Quality/projects/golang/05_schema_validation_gate/main.go
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/santhosh-tekuri/jsonschema/v5"
)

// defineSchema compiles a JSON schema from a string.
// In production, this would be loaded from a Schema Registry.
func defineSchema() (*jsonschema.Schema, error) {
	schemaStr := `{
		"$schema": "https://json-schema.org/draft/2020-12/schema",
		"title": "User Event",
		"type": "object",
		"properties": {
			"event_id": {
				"type": "string",
				"minLength": 10
			},
			"user_id": {
				"type": "integer",
				"minimum": 1
			},
			"is_active": {
				"type": "boolean"
			}
		},
		"required": ["event_id", "user_id", "is_active"]
	}`

	compiler := jsonschema.NewCompiler()
	if err := compiler.AddResource("schema.json", strings.NewReader(schemaStr)); err != nil {
		return nil, err
	}
	
	return compiler.Compile("schema.json")
}

// processEvent validates a raw JSON string against the compiled schema.
func processEvent(schema *jsonschema.Schema, rawJSON string) {
	fmt.Printf("\nProcessing payload:\n%s\n", rawJSON)

	// jsonschema expects an unmarshaled map[string]interface{} or []interface{}
	// Or we can validate directly from a string using string reader
	v, err := jsonschema.UnmarshalJSON(strings.NewReader(rawJSON))
	if err != nil {
		fmt.Printf("❌ Invalid JSON format: %v\n", err)
		return
	}

	err = schema.Validate(v)
	if err != nil {
		fmt.Println("❌ Schema Validation Failed. Routing to DLQ...")
		// Print detailed validation errors
		fmt.Printf("Validation Errors:\n%#v\n", err)
	} else {
		fmt.Println("✅ Schema Validation Passed! Proceeding to ingest event.")
	}
}

func main() {
	schema, err := defineSchema()
	if err != nil {
		log.Fatalf("Failed to compile JSON schema: %v", err)
	}

	// Good Payload
	goodPayload := `{
		"event_id": "evt_123456789",
		"user_id": 42,
		"is_active": true,
		"platform": "web"
	}`

	// Bad Payload: missing required user_id, invalid type for is_active, event_id too short
	badPayload := `{
		"event_id": "evt_123",
		"is_active": "yes"
	}`

	processEvent(schema, goodPayload)
	processEvent(schema, badPayload)
}
